### PR TITLE
[Metrics] add more metrics

### PR DESCRIPTION
**Pull request by HarryWu99**
### metrics
- time_queue_requests
- time_inference_requests
- time_prefill_requests
- time_decode_requests
- max_num_generation_tokens_requests

```
max_num_generation_tokens_requests.append(
    max(seq.get_output_len() for seq in seq_group.get_seqs())
)
```

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by HarryWu99** [comment]
@robertgshaw2-neuralmagic I just replicating the new metrics that were added [here](https://github.com/ronensc/vllm/pull/1). should I add some automated tests, or just visually look at the metrics？Thanks!

File: vllm/engine/metrics.py [hunk 1]
```diff
@@ -185,11 +191,16 @@ class Stats:
     # Request stats (should have _requests suffix)
     #   Latency
     time_e2e_requests: List[float]
+    time_queue_requests: List[float]
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/metrics.py:194]: looks like none of these are actually being logged to prometheus?

**Review by robertgshaw2-redhat — CHANGES_REQUESTED** [review]

**Comment by robertgshaw2-neuralmagic** [comment]
> @robertgshaw2-neuralmagic I just replicating the new metrics that were added [here](https://github.com/ronensc/vllm/pull/1). should I add some automated tests, or just visually look at the metrics？Thanks!

This one will be tricky to test given its a timing flow

**Comment by robertgshaw2-neuralmagic** [comment]
Also, looks like you are missing the number of tokens in the batch for each iteration of the LLMEngine

**Comment by robertgshaw2-neuralmagic** [comment]
Also, see https://github.com/vllm-project/vllm/issues/3616 for some more ideas

**Comment by HarryWu99** [comment]
> Also, looks like you are missing the number of tokens in the batch for each iteration of the LLMEngine

you mean `num_generation_tokens_iter`? It already exists.

**Comment by HarryWu99** [comment]
https://github.com/vllm-project/vllm/issues/3616#issuecomment-2030858781

@robertgshaw2-neuralmagic BTW, Should some metrics be recorded with a periodic bypass process/thread, such as every 1 second.

File: vllm/engine/metrics.py (hunk 1 above)
Comment (HarryWu99) [reply vllm/engine/metrics.py:218]: yes. sorry. I forget. Have modified

**Comment by robertgshaw2-neuralmagic** [comment]
> [#3616 (comment)](https://github.com/vllm-project/vllm/issues/3616#issuecomment-2030858781)
> 
> @robertgshaw2-neuralmagic BTW, Should some metrics be recorded with a periodic bypass process/thread, such as every 1 second.

We have this functionality already in the `StatLogger`, but I don't think any of these metrics should go through that pathway

**Comment by robertgshaw2-neuralmagic** [comment]
> > Also, looks like you are missing the number of tokens in the batch for each iteration of the LLMEngine
> 
> you mean `num_generation_tokens_iter`? It already exists.

For chunked prefill, we can have prefill and generation tokens in the same batch

So I was looking for `num_tokens_iter` which would be the sum of `num_generation_tokens_iter + num_prompt_tokens_iter`

**Comment by robertgshaw2-neuralmagic** [comment]
@HarryWu99 can you update the grafana dashboard and turn this into a real PR and not draft? would like to merge before friday (so it can be the release)

To run grafana, checkout `examples/production_monitoring` - I set up a docker compose to spin it up easily

**Comment by HarryWu99** [comment]
@robertgshaw2-neuralmagic Thanks for attention!

I made some changes to the dashboard. Categorize the charts. But I'm not sure I'm showing some metrics the right way, such as:
- request_params_n
- request_best_of
- request_max_num_generation_tokens

**Comment by robertgshaw2-neuralmagic** [comment]
ok will try it out today

**Comment by HarryWu99** [comment]
@robertgshaw2-neuralmagic Hi, will it still be merged before friday?

**Comment by robertgshaw2-neuralmagic** [comment]
> @robertgshaw2-neuralmagic Hi, will it still be merged before friday?

Working on merging this week

**Comment by HarryWu99** [comment]
@robertgshaw2-neuralmagic ping~

**Comment by robertgshaw2-neuralmagic** [comment]
sorry - traveling this week. back to PR review.

Ill fix conflict + merge tom

**Review by robertgshaw2-redhat — APPROVED** [review]

**Comment by HarryWu99** [comment]
@robertgshaw2-neuralmagic Hello~ I see it's been approved for a while, but it is not merged yet. Anything wrong?

**Comment by LucasWilkinson** [comment]
@HarryWu99 thanks for putting this PR together, Im interested in some of the metrics here too, it looks like it was approved with auto-merge enabled which means it will merge once the CI test go green, it appears however there are some bugs the CI is catching:

https://buildkite.com/vllm/ci/builds/8274#018fbcae-fbed-4edb-aad3-86d0aae11fad/51-695

**Comment by HarryWu99** [comment]
I have made some changes, but the CI still fails. I get a bit confused, the problem reported by the CI seems to have nothing to do with this PR. @LucasWilkinson

**Comment by LucasWilkinson** [comment]
@HarryWu99 thanks for updating the PR, a lot the tests can be flaky, I re-ran some of them to see if its just flakiness, although Im not familiar with the tool use test @robertgshaw2-neuralmagic may know more

you are rebased off latest main correct?

**Comment by HarryWu99** [comment]
@LucasWilkinson  Yes, I rebased my branch onto the `main` branch.

**Comment by vrdn-23** [comment]
@DarkLight1337  @HarryWu99 Just curious what exactly is the difference between `time_queue_requests` and the already existing `vllm:time_in_queue_requests`?

**Pull request merged**